Bash completion

filalex77 opened the pull request:
The easiest way to have maintainable completions was to make `tmatrix --help` output into GNU-style help by making simple formatting changes.

Tested on `zsh 5.6.2` and `bash 4.4.23`.

![zsh completion screenshot](https://i.imgur.com/xu5tViH.png)

filalex77 commented on extra/completions/_tmatrix:
```diff
@@ -0,0 +1,2 @@
+#compdef tmatrix
+_gnu_generic tmatrix
```
Thank you, I was just thinking on how to do that.

filalex77 replied:
`--` for `_arguments` is really not easy to find in `man zshcompsys` :(

eliasrg replied:
It's certainly not the most intuitive syntax...

filalex77 commented:
@eliasrg Thank you for your help!

M4444 commented on include/Parser.h:
```diff
@@ -171,8 +171,7 @@ namespace Parser {
 			{
-				"Set the range for the update rate of the",
-				"Matrix characters",
+				"Set the range for the update rate of the Matrix characters",
```
This exceeds 80 characters when help is printed. I'm flexible on other things but help must fit inside a width of 80 characters.

filalex77 replied:
Okay, I hear that. Then it's probably best to reword this?

filalex77 replied:
How about something like `Set the range for the character update rate`? It's informative enough, and fits.

M4444 replied:
Yup, that works. I that case also remove `Matrix` from the description of the color option.

filalex77 replied:
Done.

M4444 commented on src/Parser.cpp:
```diff
@@ -21,19 +21,19 @@ namespace Parser {
 			break;
 		case RANGE:
-			return "<min>,<max>";
+			return "RANGE";
```
The current style is largely based on git. With this change it might even be prettier but it's unintuitive how RANGE is supposed to be represented. Do you have an examples of other applications that have arguments that look similar to this, especially arguments that need 2 numbers like RANGE?

filalex77 replied:
I went looking, but no, `RANGE`-style arguments are not really used. Though, many programs, ones by GNU in particular, use parameter names like `DELIMITER` all over the place.

I think that it's the job of the manual to explain the semantics of `RANGE` and other terms, and `--help` text should not focus on that.

M4444 replied:
Can you give me examples programs use `DELIMITER`? Without looking it up I don't understand what it is or how it relates to `RANGE`.

Why do you think `--help` shouldn't focus on explaining the semantics of arguments? Do you have examples of programs that don't? A counterexample is `ls --help`, which explains the semantics of all the arguments.

filalex77 replied:
Nevermind - `DELIMITER` is a simple string. Used by `gawk`, for example. Not related, sorry.

You're right - `ls` does that with `SIZE`. Then it's probably best to do the same - move the syntax description for ranges from the option descriptions, and into a separate block below them.

M4444 replied:
Yea, we could add something like:
```
RANGE is a pair of numbers separated by a comma, where the first number should be less or equal to the second number. It specifies the boundaries of an integer set which will be used in random operations.
```

M4444 replied:
Also since this is now changing let's add a space after the short arguments in the help menu.
```
diff --git a/src/Parser.cpp b/src/Parser.cpp
index 34076f8..c58fc23 100644
--- a/src/Parser.cpp
+++ b/src/Parser.cpp
@@ -53,10 +53,10 @@ namespace Parser {
                case COLOR:
                case TEXT:
                        if (ShortLiteral != "" && LongLiteral != "") {
-                               return "[" + ShortLiteral + valueName + " | " +
+                               return "[" + ShortLiteral + " " + valueName + " | " +
                                       LongLiteral + "=" + valueName + "]";
                        } else if (ShortLiteral != "") {
-                               return "[" + ShortLiteral + valueName + "]";
+                               return "[" + ShortLiteral + " " + valueName + "]";
                        } else {
                                return "[" + LongLiteral + "=" + valueName + "]";
                        }
```

M4444 commented:
I didn't yet had time to test the completions but I just wanted to clarify what is expected of them. Git is a good benchmark because it has many flags and pretty good completions. I've messed around with bash completions and it's not that hard to implement short options like `-C` but long options that end in `=` like `--color=` become a bit tricky and probably require custom functions.
Here's a test example - when you type:
```
git show --pr
```
and hit `Tab` you get:
```
git show --pretty=
```
Notice the cursor is __just after__ `=`, __no empty space was added__.

A couple more tabs and the options appear:
```
email     format:   full      fuller    medium    oneline   raw       short
```
\
The completions for TMatrix should work in a similar way, for example for the `--color=` flag.

eliasrg commented:
> I didn't yet had time to test the completions but I just wanted to clarify what is expected of them. Git is a good benchmark because it has many flags and pretty good completions. I've messed around with bash completions and it's not that hard to implement short options like `-C` but long options that end in `=` like `--color=` become a bit tricky and probably require custom functions.
> Here's a test example - when you type:
> 
> ```
> git show --pr
> ```
> 
> and hit `Tab` you get:
> 
> ```
> git show --pretty=
> ```
> 
> Notice the cursor is **just after** `=`, **no empty space was added**.
> 
> A couple more tabs and the options appear:
> 
> ```
> email     format:   full      fuller    medium    oneline   raw       short
> ```
> 
> 
> The completions for TMatrix should work in a similar way, for example for the `--color=` flag.

They do work this way, at least the zsh completions. I'm not sure how to use the bash ones.

filalex77 commented:
@M4444 Currently, these `bash` completions do about 50% of what you described. `zsh` makes it easy to add values for such arguments, but I'm currently not sure on how to easily do that in `bash`. Completions for `git` are huge...

Here's what I have now (my shell is `zsh`): https://asciinema.org/a/9qkc4RC65PTIvE1KIhZc4cAYc

filalex77 commented:
@M4444 I did an alpha-grade `bash` completion in bcd239a, could you please take a look? Now it completes modes and color names, and doesn't put unnecessary spaces.

filalex77 commented:
@M4444 If you are fine with `RANGE`, this can probably be merged.

Also, we can do a separate PR for changes to the `man` page like these.

M4444 commented:
> Also, we can do a separate PR for changes to the `man` page like these.

I don't want to disjoin the man page from the rest, so if this gets merged there will be 2 commits. The first one will contain the changes to the source and the equivalent changes in the man page, while the second will have the completions. If the bash completions aren't yet ready they can be separated into a third commit.

filalex77 commented:
> > Also, we can do a separate PR for changes to the `man` page like these.
> 
> I don't want to disjoin the man page from the rest, so if this gets merged there will be 2 commits. The first one will contain the changes to the source and the equivalent changes in the man page, while the second will have the completions. If the bash completions aren't yet ready they can be separated into a third commit.

Okay, I'll do a rebase now.

M4444 commented:
> @M4444 I did an alpha-grade `bash` completion in [bcd239a](https://github.com/M4444/TMatrix/commit/bcd239abf49b46a7577e5206a40fcf0caae98bcf), could you please take a look? Now it completes modes and color names, and doesn't put unnecessary spaces.

The bash completion is looking good. Completion of short arguments still need to be added so lets separate this into a third commit so the other two can be merged quicker.

M4444 commented on tmatrix.6:
```diff
@@ -58,7 +58,7 @@ The minimal length value is \fB1\fR.
 .TP 4
 \-r <min>,<max>
-Set the range for the update rate of the Matrix characters.
+Set the range for the character update rate.
```
This is a good start but the whole man paged should be changed to reflect the changes in the help menu.

filalex77 replied:
I updated the parameter names. Probably can work on different things later.

M4444 commented on CMakeLists.txt:
```diff
@@ -38,3 +38,10 @@ add_executable(tmatrix ${tmatrix_SRC})
 install(TARGETS tmatrix DESTINATION bin)
+
+# Completions
+install(FILES extra/completions/_tmatrix
```
I think we can name them tmatrix-completion.bash and tmatrix-completion.zsh and then renamed them appropriately in the final directory. The repo directory can just be `completion` instead of `extra/completions`.
Move the closing parentheses into the next row.

M4444 commented on extra/completions/tmatrix.bash:
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
```
Completion script doesn't need this but it will probably need a licensing header.

filalex77 commented:
@M4444 Thank you for all the feedback; I'll be on it soon.

filalex77 commented:
@M4444 I did some work on the changes you requested; feel free to review when you have time.

M4444 commented on tmatrix.6:
```diff
@@ -3,13 +3,13 @@
-        [-r<min>,<max>] [-C<color> | --color=<color>]
-        [-c<color> | --background=<color>] [-t<text> | --title=<text>]
+\fItmatrix\fR [--version] [--help] [--mode=MODE]
+        [-sVALUE|--steps-per-sec=VALUE]
```
Also add the spaces here, e.g.:
```suggestion
        [-s VALUE|--steps-per-sec=VALUE]
```

M4444 commented on completions/tmatrix-completion.bash:
```diff
@@ -0,0 +1,43 @@
+#
```
No need for a full license header, just add the shortened SPDX-License-Identifier version like it is in the source files.

filalex77 replied:
Sorry, didn't pay attention to that.

M4444 commented on CMakeLists.txt:
```diff
@@ -48,3 +48,13 @@ add_custom_target(manpage ALL DEPENDS tmatrix.6.gz)
 install(TARGETS tmatrix DESTINATION bin)
 install(FILES ${PROJECT_BINARY_DIR}/tmatrix.6.gz DESTINATION share/man/man6)
+
+# Completions
```
`Completions` -> `Install bash and zsh completion scripts.`

Also swap the order of the zsh and bash install commands.

filalex77 replied:
Sure.

M4444 commented:
> 
> EDIT: sorry for an awful lot of force pushes; I mixed up my `master` and yours when rebasing.

It's not a problem. It's always better when everything is correctly rebased to master. You can squash the commits into the 3 I described before.

filalex77 commented:
> You can squash the commits into the 3 I described before.

Let me make this clear: we should have 3 commits - 
- Changed output of --help and manpage
- Added completions for zsh
- Added completions for bash

Where bash completions can be worked on in a separate PR.

Is this right?

M4444 commented:
Yes, that's right. I'll merge the first 2 commits now and I'll leave it up to you if you want to separate bash completions into a new PR so this one can be merged and closed now or you want bash completions to stay here in which case the the PR would remain open unit they are done.

filalex77 commented:
> or you want bash completions to stay here in which case the the PR would remain open

Sure, I'd like to finish with short arguments before everything is merged.

filalex77 commented:
I've looked into bash completions some more, mostly into the examples from [scop/bash-completion](https://github.com/scop/bash-completion).

I think that bash doesn't provide for parameter naming like zsh. And `_parse_help` doesn't parse short arguments for options that have the long versions.

For now, I hardcoded the list of short options into the completion script, tagged it as `FIXME`. Let me know if you have any ideas.

M4444 commented:
Merged the first 2 commits with some changes:
- Added a description of RANGE
- Adapted the manpage and zsh script to accurately mirror the current --help output
- Reworded the commit messages a bit

M4444 commented:
> 
> For now, I hardcoded the list of short options into the completion script, tagged it as `FIXME`. Let me know if you have any ideas.

I don't any better ides so I think it's fine to commit it like this. You can remove the `FIXME`. If anyone familiar with this PR comes up with a better way to do it in the future they will know what to fix.

Also, use tabs for indentation to conform with the rest of the project.

filalex77 commented:
@M4444 Done, now we have only bash completion here.

M4444 commented on completions/tmatrix-completion.bash:
```diff
@@ -0,0 +1,35 @@
+#
```
Remove the 1st and 5th line, no need to expand the header.

M4444 commented on CMakeLists.txt:
```diff
@@ -49,7 +49,11 @@ add_custom_target(manpage ALL DEPENDS tmatrix.6.gz)
-# Install zsh completion script.
+# Install bash and zsh completion scripts.
+install(FILES completions/tmatrix-completion.bash
+        DESTINATION share/bash-completion/completions
```
Tabs here too.

M4444 commented:
Just leave the commit with the title, the commit message was probably left over from previous commits.

filalex77 commented:
@M4444 Thank you for the continued help. Hope it's good now.

M4444 commented:
@filalex77 thanks for you countribution and @eliasrg thank you for the suggestions and review. I think the completion scripts turned out really great and I really appreciate your unwavering effort.

eliasrg commented:
@M4444 My pleasure. It was interesting to learn how completions work!
I can also confirm that the Arch package builds without changes.

filalex77 commented:
@eliasrg @M4444 Yes, this is really cool. Glad to have completions on my Gentoo :)

M4444 commented:
Great, it's nice to know that everything works fine for your distributions.
